[Cocoa, MediaStream] Support AVCapture virtual cameras

eric-carlson opened the pull request:
#### ba4235d06159cea7784a8c6962a9facd2e0c8a9b
<pre>
[Cocoa, MediaStream] Support AVCapture virtual cameras
<a href="https://bugs.webkit.org/show_bug.cgi?id=250142">https://bugs.webkit.org/show_bug.cgi?id=250142</a>
rdar://103925476

Reviewed by Youenn Fablet.

Add support for AVCapture virtual cameras. When getUserMedia video constraints request
`facingMode : environment`, prefer the wide angle camera with the greatest capabilities.

* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm: Include additional camera constants.

* Source/WebCore/platform/mediastream/RealtimeMediaSource.h: Make `fitnessDistance` virtual.
* Source/WebCore/platform/mediastream/RealtimeVideoSource.h: Implement `fitnessDistance`,
pass through to source.

* Source/WebCore/platform/mediastream/mac/AVCaptureDeviceManager.mm:
(WebCore::cameraCaptureDeviceTypes): Include AVCaptureDeviceTypeBuiltInDualWideCamera and
AVCaptureDeviceTypeBuiltInTripleCamera in types array.

* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::fitnessDistance): Add facingModeFitnessDistanceAdjustment to
calculated `facingMode` fitness distance.

* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::facingModeFitnessDistanceAdjustment const): Return a value to
prefer wide angle cameras, and so when a device has more than one wide angle camera we prefer
the wide angle camera with the greatest capabilities.

Canonical link: <a href="https://commits.webkit.org/258585@main">https://commits.webkit.org/258585@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/564ebc4b0224f0ba028c55385aec700c2e89b719

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102341 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11479 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35408 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111650 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171822 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106317 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12480 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2398 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94669 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109359 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108122 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9551 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37291 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24302 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/79038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4991 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25725 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5130 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2165 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11160 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45218 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5894 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6882 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
```diff
@@ -242,6 +242,8 @@ class WEBCORE_EXPORT RealtimeMediaSource
     const CaptureDevice& captureDevice() const { return m_device; }
     bool isEphemeral() const { return m_device.isEphemeral(); }
 
+    virtual double fitnessDistance(const MediaConstraint&);
```
How about leaving fitnessDistance non virtual but adding a virtual computeFacingModeFitnessDistance?

eric-carlson replied:
Good idea!

youennf reviewed (approved):
LGTM

eric-carlson committed:
[Cocoa, MediaStream] Support AVCapture virtual cameras
https://bugs.webkit.org/show_bug.cgi?id=250142
rdar://103925476

Reviewed by Youenn Fablet.

Add support for AVCapture virtual cameras. When getUserMedia video constraints request
`facingMode : environment`, prefer the wide angle camera with the greatest capabilities.

* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm: Include additional camera constants.

* Source/WebCore/platform/mediastream/RealtimeMediaSource.h: Make `fitnessDistance` virtual.
* Source/WebCore/platform/mediastream/RealtimeVideoSource.h: Implement `fitnessDistance`,
pass through to source.

* Source/WebCore/platform/mediastream/mac/AVCaptureDeviceManager.mm:
(WebCore::cameraCaptureDeviceTypes): Include AVCaptureDeviceTypeBuiltInDualWideCamera and
AVCaptureDeviceTypeBuiltInTripleCamera in types array.

* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::fitnessDistance): Add facingModeFitnessDistanceAdjustment to
calculated `facingMode` fitness distance.

* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::AVVideoCaptureSource::facingModeFitnessDistanceAdjustment const): Return a value to
prefer wide angle cameras, and so when a device has more than one wide angle camera we prefer
the wide angle camera with the greatest capabilities.

Canonical link: https://commits.webkit.org/258585@main
